tests/python_package_test/test_basic.py

```diff
@@ -39,6 +39,8 @@ def test(self):
         self.assertEqual(bst.current_iteration(), 20)
         self.assertEqual(bst.num_trees(), 20)
         self.assertEqual(bst.num_model_per_iteration(), 1)
+        self.assertAlmostEqual(bst.upper_bound(), 3.32, places=2)
+        self.assertAlmostEqual(bst.lower_bound(), -3.13, places=2)
```

Comment:
`places=2` seems to be very poor comparison. Do you have any thoughts why is it fail with more strict checks?

Reply:
I agree it is really poor, I do not know what is the reason, may be something about the interaction and transformation with Python? I based my basic.py implementation on get_leaf_output which is not tested, and I wonder if it might also have some dependency on platforms? I am not sure how python transforms c types. May you have a clue about this?

Reply:
Or could it be that the trained model has slight variations? maybe it would be best to only test from a precomputed model that it gets the value it is supposed to get with less interactions